src/tuner/nccl_ofi_tuner.c

```diff
@@ -0,0 +1,107 @@
+#include <stdlib.h>
+#include "nccl-headers/tuner.h"
+#include "nccl_ofi_brasso.h"
+#include "nccl_ofi_log.h"
+
+ncclResult_t brasso_init(size_t nRanks, size_t nNodes, ncclDebugLogger_t logFunction)
+{
+	ofi_log_function = logFunction;
+
+	if (getenv("NCCL_ALGO") || getenv("NCCL_PROTO")) {
+		NCCL_OFI_WARN("The tuner plugin can not be loaded when explicitly choosing an algorithm or protocol with NCCL_ALGO/NCCL_PROTO");
+		return ncclInvalidUsage;
+	}
+
+	/* TODO: Make this tunable with envs or grab from config  */
+	struct brasso_model_params params = {
+		.net_lat = NET_LATENCY,
+		.internode_bw = INTRANODE_BW,
+		.intranode_bw = INTERNODE_BW,
+		.rails = NET_NUM_RAILS
+	};
+
+	ctx = calloc(1, sizeof(struct brasso_context));
+	if (!ctx) {
+		NCCL_OFI_WARN("Context allocation failed.");
+		return ncclInternalError;
+	}
+
+	ctx->num_ranks = nRanks;
+	ctx->num_nodes = nNodes;
+	ctx->model_params = params;
+
+	/*
+	 * Build cost model to use from brasso_get_coll_info.
+	 */
+	brasso_model_costs();
+
+	NCCL_OFI_TRACE(NCCL_TUNING, "Tuner init: comm with %ld ranks and %ld nodes.", nRanks, nNodes);
+	return ncclSuccess;
+}
+
+ncclResult_t brasso_get_coll_info(ncclFunc_t collType, size_t nBytes,
+				  int collNetSupport, int nvlsSupport, int numPipeOps,
+				  int *algorithm, int *protocol, int* nChannels)
+{
+	float cost = 0;
+	float lowest = FLT_MAX;
+	int algo, proto = 0;
+
+	/* Skip runs smaller than 2 nodes */
+	if (ctx->num_nodes <= 2)
+		return ncclSuccess;
+
+	/*
+	 * Ideally, this should just be a lookup and not be in-flight math
+	 * We do not want divs in the hot path, but working with the API we've
+	 * got now. 
+	 */
+	for (algo = 0; algo < NCCL_NUM_ALGORITHMS; algo++) {
+		/* No CollNet on AWS today */
+		if (algo == NCCL_ALGO_COLLNET_DIRECT || algo == NCCL_ALGO_COLLNET_CHAIN)
+			continue;
+
+		/* No IB Sharp on AWS, so skip NCCL_ALGO_NVLS */
```

Comment:
`NCCL_ALGO_NVLS` just mean using NVLS intra-node on single node. It isn't related to IB-SHARP which is `NCCL_ALGO_COLLNET_DIRECT` and `NCCL_ALGO_COLLNET_CHAIN` above. You should skip it but comment is wrong.